### PR TITLE
feat(core): make ValidatorRegistry extensible with plugin architecture

### LINKED ISSUE #348: Make ValidatorRegistry extensible with plugin architecture
## Context
From code audit (architecture review). The `ValidatorRegistry` has 19 hardcoded validator factories in `lib.rs`.

## Current State
- Validators are registered via hardcoded factory functions in `lib.rs:155-203`
- No way to disable specific validators at runtime (only rules)
- Adding a 20th validator requires editing lib.rs

## Proposal
1. Add `ValidatorProvider` trait for plugin-like registration
2. Use builder pattern to make registry immutable after setup
3. Add `validator_enabled()` check in `validators_for()`
4. Group validators into sub-modules (builtin, tools, generic)

## Location
`crates/agnix-core/src/lib.rs:121-204`

## Effort
Medium - refactoring but maintains backward compat

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

Adds a plugin architecture to `ValidatorRegistry` enabling external validator registration and runtime validator control. Closes #348.

- **`ValidatorProvider` trait** -- Extension point for supplying validator factories from external sources (plugins, secondary rule sets)
- **`ValidatorRegistryBuilder`** -- Builder pattern for ergonomic registry construction with `with_defaults()`, `with_provider()`, `without_validator()`, `build()`
- **`Validator::name()`** -- Default method returning the short type name for programmatic identification
- **`disabled_validators` config** -- New `[rules]` field to disable validators by name at runtime
- **Performance optimization** -- `validators_for()` skips filtering entirely when no validators are disabled (zero overhead for common case)

## Key Files

| File | Change |
|------|--------|
| `crates/agnix-core/src/registry.rs` | ValidatorProvider, BuiltinProvider, ValidatorRegistryBuilder, disabled filtering |
| `crates/agnix-core/src/rules/mod.rs` | Validator::name() with default impl |
| `crates/agnix-core/src/config.rs` | disabled_validators field on RuleConfig |
| `crates/agnix-core/src/pipeline.rs` | Wire disabled_validators into convenience functions |
| `crates/agnix-core/src/lib.rs` | Re-export new public types |

## Backward Compatibility

All existing public APIs unchanged:
- `ValidatorRegistry::new()`, `with_defaults()`, `register()`, `validators_for()`, `Default` impl
- `ValidatorFactory` type alias unchanged (`fn() -> Box<dyn Validator>`)
- `Validator` trait gains `name()` with default impl (no changes to existing implementors)
- `RuleConfig` gains `disabled_validators` with `#[serde(default)]` (existing TOML configs parse fine)

## Test Plan

- [x] 2660+ tests pass (including ~40 new tests for plugin architecture)
- [x] API contract tests verify new types are importable and compile
- [x] Send+Sync bounds verified for ValidatorRegistryBuilder
- [x] Cross-crate integration tests verify external usage
- [x] Edge cases: multiple disabled validators, all validators disabled, idempotent disabling, multiple providers
- [x] TOML deserialization tests for disabled_validators config field
- [x] Clippy clean, format clean, release build passes

Closes #348

### CODE REVIEW

File: crates/agnix-core/src/registry.rs [hunk 1]
```diff
@@ -113,6 +165,130 @@ impl Default for ValidatorRegistry {
...
+        Self {
+            entries: Vec::new(),
+            disabled_validators: HashSet::new(),
+        }
+    }
+
+    /// Add all built-in validators (equivalent to [`ValidatorRegistry::with_defaults`]).
+    pub fn with_defaults(&mut self) -> &mut Self {
+        self.with_provider(&BuiltinProvider)
+    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/registry.rs:207]: Fixed in 1581311. Added doc comment noting with_defaults() is additive and warning against duplicate calls.

File: CLAUDE.md [hunk 2]
```diff
@@ -72,19 +72,29 @@ tests/fixtures/     # Test cases by category
...
-- `registry.rs` - ValidatorRegistry, factory functions
+- `registry.rs` - ValidatorRegistry, ValidatorRegistryBuilder, ValidatorProvider, factory functions
 
 ### Key Abstractions
 
 ```rust
 // Primary extension point
 pub trait Validator {
     fn validate(&self, path: &Path, content: &str, config: &LintConfig) -> Vec<Diagnostic>;
+    fn name(&self) -> &'static str; // Defaults to short type name
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply CLAUDE.md]: Fixed in 1581311. Updated snippet to show default body syntax: `fn name(&self) -> &'static str { /* default: short type name */ }`

File: CLAUDE.md [hunk 3]
```diff
@@ -72,19 +72,29 @@ tests/fixtures/     # Test cases by category
...
 }
 
-// Multiple validators per FileType via factory pattern
-pub struct ValidatorRegistry {
-    validators: HashMap<FileType, Vec<ValidatorFactory>>
+// Plugin architecture for extensibility
+pub trait ValidatorProvider {
+    fn validators(&self) -> Vec<(FileType, ValidatorFactory)>;
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply CLAUDE.md]: Fixed in 1581311. Updated snippet to show Send + Sync bounds and default name() method.

File: AGENTS.md [hunk 4]
```diff
@@ -72,19 +72,29 @@ tests/fixtures/     # Test cases by category
...
-- `registry.rs` - ValidatorRegistry, factory functions
+- `registry.rs` - ValidatorRegistry, ValidatorRegistryBuilder, ValidatorProvider, factory functions
 
 ### Key Abstractions
 
 ```rust
 // Primary extension point
 pub trait Validator {
     fn validate(&self, path: &Path, content: &str, config: &LintConfig) -> Vec<Diagnostic>;
+    fn name(&self) -> &'static str; // Defaults to short type name
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply AGENTS.md]: Fixed in 1581311. AGENTS.md synced with CLAUDE.md (byte-for-byte identical).

File: AGENTS.md [hunk 5]
```diff
@@ -72,19 +72,29 @@ tests/fixtures/     # Test cases by category
...
 }
 
-// Multiple validators per FileType via factory pattern
-pub struct ValidatorRegistry {
-    validators: HashMap<FileType, Vec<ValidatorFactory>>
+// Plugin architecture for extensibility
+pub trait ValidatorProvider {
+    fn validators(&self) -> Vec<(FileType, ValidatorFactory)>;
+}
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply AGENTS.md:90]: Fixed in 1581311. AGENTS.md synced with CLAUDE.md.

File: crates/agnix-core/src/registry.rs [hunk 6]
```diff
@@ -113,6 +165,130 @@ impl Default for ValidatorRegistry {
...
+        self
+    }
+
+    /// Consume the builder and produce a [`ValidatorRegistry`].
+    pub fn build(&mut self) -> ValidatorRegistry {
+        let mut registry = ValidatorRegistry {
+            validators: HashMap::new(),
+            disabled_validators: std::mem::take(&mut self.disabled_validators),
+        };
+        for &(file_type, factory) in &self.entries {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/registry.rs:241]: Fixed in 1581311. Updated doc comment to accurately describe the drain semantics: build() uses mem::take on the disabled set, so subsequent calls produce registries without disabled validators. The entries (factory registrations) are preserved since they use reference iteration.

File: crates/agnix-core/src/rules/mod.rs [hunk 7]
```diff
@@ -23,9 +23,41 @@ pub mod xml;
...
-/// Trait for file validators
-pub trait Validator {
+/// Extract the short (unqualified) type name from `std::any::type_name`.
+///
+/// Given a fully-qualified path like `"agnix_core::rules::skill::SkillValidator"`,
+/// returns `"SkillValidator"`. Falls back to the full name when no `::` separator
+/// is found.
+fn short_type_name<T: ?Sized + 'static>() -> &'static str {
+    let full = std::any::type_name::<T>();
+    full.rsplit("::").next().unwrap_or(full)
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/mod.rs]: Fixed in 1581311. Updated short_type_name() to strip generic suffix before splitting on `::'. Now handles `Wrapper<foo::Bar>` correctly by splitting on '<' first, yielding 'Wrapper' instead of 'Bar>'.

File: crates/agnix-core/src/registry.rs [hunk 8]
```diff
@@ -47,60 +117,42 @@ impl ValidatorRegistry {
...
+
+        if self.disabled_validators.is_empty() {
+            return factories.iter().map(|factory| factory()).collect();
+        }
+
+        factories
+            .iter()
             .map(|factory| factory())
+            .filter(|v| !self.disabled_validators.contains(v.name()))
             .collect()
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/registry.rs:138]: Thanks for the positive feedback. The early-return optimization was added specifically for this hot path.
